src/Features/Core/Portable/Snippets/RoslynLSPSnippetConverter.cs

```diff
@@ -0,0 +1,161 @@
+            // character position.
+            // If there is a caret at the end of the line, then it's position
+            // will be equivalent to the length of the TextChange.
+            for (var i = 0; i < textChangeText.Length + 1;)
+            {
+                if (i == caretPosition - textChangeStart)
+                {
+                    lspSnippetString.Append("$0");
+
+                    // Special case for cursor position since they will occur between positions
+                    // so we still wants to insert the character following the cursor position.
+                    // Will not happen for placeholders since they have a direct mapping from position
+                    // of the identifier to their position in the TextChange text.
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                    }
+
+                    i++;
+                }
+
+                //Tries to see if a value exists at that position in the map, and if so it
+                // generates a string that is LSP formatted.
+                if (dictionary.TryGetValue(i, out var placeholderInfo))
+                {
+                    var str = $"${{{placeholderInfo.priority}:{placeholderInfo.identifier}}}";
+                    lspSnippetString.Append(str);
+
+                    // Skip past the entire identifier in the TextChange text
+                    i += placeholderInfo.identifier.Length;
+                }
+                else
+                {
+                    if (i < textChangeText.Length)
+                    {
+                        lspSnippetString.Append(textChangeText[i]);
+                        i++;
+                    }
+                    else
+                    {
+                        break;
+                    }
+                }
+            }
+
+            return lspSnippetString.ToString();
+        }
+
+        /// <summary>
+        /// Preprocesses the list of placeholders into a dictionary that maps the insertion position
+        /// in the string to the placeholder's identifier and the priority associated with it.
+        /// </summary>
+        private static void GetMapOfSpanStartsToLSPStringItem(ref PooledDictionary<int, (string identifier, int priority)> dictionary, ImmutableArray<SnippetPlaceholder> placeholders, int textChangeStart)
+        {
+            for (var i = 0; i < placeholders.Length; i++)
+            {
+                var placeholder = placeholders[i];
+                foreach (var position in placeholder.PlaceHolderPositions)
+                {
+                    // i + 1 since the placeholder priority is set according to the index in the
+                    // placeholders array, starting at 1.
+                    dictionary.Add(position - textChangeStart, (placeholder.Identifier, i + 1));
```

Comment:
what if you have two placeholders at the same position.  this will throw on .Add?